File: kiko-api/src/repositories/judgeRepository.ts
/**
 * Judge Decision Repository
 * Saves Judge Engine v3.5 decisions to database for ML/DL training
 */

import { prisma, withRetry } from '../db/prisma.js';
import { DecisionEngineOutput } from '../types/judgeTypes.js';

/**
 * Save a judge decision to database
 */
export async function saveJudgeDecision(
    judgeOutput: DecisionEngineOutput,
    analysisTimeMs: number
): Promise<string> {
    const engine = judgeOutput.decision_engine;

    const decision = await withRetry(() => prisma.judgeDecision.create({
        data: {
            // Input
            tokenAddress: engine.input.token_address,
            tokenSymbol: undefined,  // TODO: Extract from token data
            tokenName: undefined,
            chainId: engine.input.chain_id,
            userAmountUsd: engine.input.user_amount,
            targetWallet: engine.input.target_wallet,

            // Context
            launchpadType: engine.input.launchpad_type,
            liquidity: engine.layers.liquidity_layer.lp_depth_usd,
            contractAgeHours: engine.layers.stage_layer.contract_age_minutes / 60,

            // Layer Outputs
            userSizeLayer: engine.layers.user_size_layer as any,
            liquidityLayer: engine.layers.liquidity_layer as any,
            structureLayer: engine.layers.structure_layer as any,
            stageLayer: engine.layers.stage_layer as any,
            tokenIntelLayer: engine.layers.token_intelligence_layer as any,

            // Final Decision
            finalDecision: engine.final_decision.decision,
            overallRiskScore: engine.final_decision.overall_risk_score,
            slippageEstimate: engine.final_decision.slippage_estimate,
            reasons: engine.final_decision.reasons as any,
            aiRationale: engine.final_decision.ai_rationale,

            // Full Output
            fullOutputJson: JSON.stringify(judgeOutput, null, 2),

            // Metadata
            analysisTimeMs,
        },
    }));

    console.log(`[Judge Repository] Saved decision ${decision.id} for ${engine.input.token_address}`);

    return decision.id;
}

/**
 * Update actual outcome after trade execution
 */
export async function updateJudgeOutcome(
    decisionId: string,
    outcome: {
        actualExecuted: boolean;
        actualProfitPct?: number;
        actualOutcome?: 'success' | 'failed' | 'rug' | 'honeypot';
        outcomeNotes?: string;
    }
): Promise<void> {
    await withRetry(() => prisma.judgeDecision.update({
        where: { id: decisionId },
        data: {
            actualExecuted: outcome.actualExecuted,
            actualProfitPct: outcome.actualProfitPct,
            actualOutcome: outcome.actualOutcome,
            outcomeNotes: outcome.outcomeNotes,
            outcomeUpdatedAt: new Date(),
        },
    }));

    console.log(`[Judge Repository] Updated outcome for decision ${decisionId}: ${outcome.actualOutcome}`);
}

/**
 * Get all judge decisions for ML training
 */
export async function getJudgeDecisionsForTraining(
    filters?: {
        chainId?: number;
        minLiquidity?: number;
        finalDecision?: string;
        hasOutcome?: boolean;
    }
): Promise<any[]> {
    const where: any = {};

    if (filters?.chainId) where.chainId = filters.chainId;
    if (filters?.minLiquidity) where.liquidity = { gte: filters.minLiquidity };
    if (filters?.finalDecision) where.finalDecision = filters.finalDecision;
    if (filters?.hasOutcome !== undefined) {
        where.actualExecuted = filters.hasOutcome ? { not: null } : null;
    }

    return prisma.judgeDecision.findMany({
        where,
        orderBy: { createdAt: 'desc' },
    });
}

/**
 * Get statistics for model evaluation
 */
export async function getJudgeStats(): Promise<any> {
    const [
        total,
        allowCount,
        allowWithRiskCount,
        blockCount,
        executedCount,
        successCount,
    ] = await Promise.all([
        prisma.judgeDecision.count(),
        prisma.judgeDecision.count({ where: { finalDecision: 'ALLOW' } }),
        prisma.judgeDecision.count({ where: { finalDecision: 'ALLOW_WITH_RISK' } }),
        prisma.judgeDecision.count({ where: { finalDecision: 'BLOCK' } }),
        prisma.judgeDecision.count({ where: { actualExecuted: true } }),
        prisma.judgeDecision.count({ where: { actualOutcome: 'success' } }),
    ]);

    return {
        total,
        decisions: {
            allow: allowCount,
            allowWithRisk: allowWithRiskCount,
            block: blockCount,
        },
        outcomes: {
            executed: executedCount,
            success: successCount,
            successRate: executedCount > 0 ? (successCount / executedCount) * 100 : 0,
        },
    };
}

File: kikoapi/internal/repositories/judge_repository.go
// Package repositories: Judge decisions (from kiko-api repositories/judgeRepository.ts).

package repositories

import (
	"context"
	"encoding/json"

	"gorm.io/gorm"

	"kikoapi/internal/types"
)

// JudgeRepository handles JudgeDecision table.
type JudgeRepository struct {
	db *gorm.DB
}

// NewJudgeRepository creates a JudgeRepository.
func NewJudgeRepository(database *gorm.DB) *JudgeRepository {
	return &JudgeRepository{db: database}
}

// SaveJudgeDecision inserts a JudgeDecision from DecisionEngineOutput.
func (r *JudgeRepository) SaveJudgeDecision(ctx context.Context, output *types.DecisionEngineOutput, analysisTimeMs int) (string, error) {
	eng := &output.DecisionEngine
	input := eng.Input
	layers := eng.Layers
	fin := eng.FinalDecision
	fullJSON, _ := json.Marshal(output)
	var id string
	err := r.db.WithContext(ctx).Raw(
		`INSERT INTO "JudgeDecision" (id, "tokenAddress", "tokenSymbol", "tokenName", "chainId", "userAmountUsd", "targetWallet", "launchpadType", liquidity, "contractAgeHours", "userSizeLayer", "liquidityLayer", "structureLayer", "stageLayer", "tokenIntelLayer", "finalDecision", "overallRiskScore", "slippageEstimate", reasons, "aiRationale", "fullOutputJson", "analysisTimeMs")
		 VALUES (gen_random_uuid()::text, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?::jsonb, ?::jsonb, ?::jsonb, ?::jsonb, ?::jsonb, ?, ?, ?, ?::jsonb, ?, ?, ?)
		 RETURNING id`,
		input.TokenAddress, nil, nil, input.ChainID, input.UserAmount, input.TargetWallet, input.LaunchpadType,
		layers.LiquidityLayer.LpDepthUSD, float64(layers.StageLayer.ContractAgeMins)/60,
		mustJSON(layers.UserSizeLayer), mustJSON(layers.LiquidityLayer), mustJSON(layers.StructureLayer), mustJSON(layers.StageLayer), mustJSON(layers.TokenIntelligenceLayer),
		string(fin.Decision), fin.OverallRiskScore, fin.SlippageEstimate, mustJSON(fin.Reasons), fin.AiRationale, string(fullJSON), analysisTimeMs,
	).Scan(&id).Error
	return id, err
}

func mustJSON(v interface{}) string {
	b, _ := json.Marshal(v)
	return string(b)
}

// UpdateJudgeOutcome updates actualExecuted, actualProfitPct, actualOutcome, outcomeNotes.
func (r *JudgeRepository) UpdateJudgeOutcome(ctx context.Context, decisionID string, actualExecuted bool, actualProfitPct *float64, actualOutcome, outcomeNotes *string) error {
	return r.db.WithContext(ctx).Exec(
		`UPDATE "JudgeDecision" SET "actualExecuted" = ?, "actualProfitPct" = ?, "actualOutcome" = ?, "outcomeNotes" = ?, "outcomeUpdatedAt" = NOW() WHERE id = ?`,
		actualExecuted, actualProfitPct, actualOutcome, outcomeNotes, decisionID,
	).Error
}

// JudgeDecisionRow for GetJudgeDecisionsForTraining.
type JudgeDecisionRow struct {
	ID string `gorm:"column:id"`
}

// GetJudgeDecisionsForTraining returns decision rows with optional filters.
func (r *JudgeRepository) GetJudgeDecisionsForTraining(ctx context.Context, chainID *int, minLiquidity *float64, finalDecision *string, hasOutcome *bool) ([]JudgeDecisionRow, error) {
	query := `SELECT id FROM "JudgeDecision" WHERE 1=1`
	args := []interface{}{}
	if chainID != nil {
		query += ` AND "chainId" = ?`
		args = append(args, *chainID)
	}
	if minLiquidity != nil {
		query += ` AND liquidity >= ?`
		args = append(args, *minLiquidity)
	}
	if finalDecision != nil {
		query += ` AND "finalDecision" = ?`
		args = append(args, *finalDecision)
	}
	if hasOutcome != nil {
		if *hasOutcome {
			query += ` AND "actualExecuted" IS NOT NULL`
		} else {
			query += ` AND "actualExecuted" IS NULL`
		}
	}
	query += ` ORDER BY "createdAt" DESC`
	var out []JudgeDecisionRow
	err := r.db.WithContext(ctx).Raw(query, args...).Scan(&out).Error
	return out, err
}

// JudgeStats for GetJudgeStats.
type JudgeStats struct {
	Total     int64
	Allow     int64
	AllowRisk int64
	Block     int64
	Executed  int64
	Success   int64
	SuccessRate float64
}

// GetJudgeStats returns counts for evaluation.
func (r *JudgeRepository) GetJudgeStats(ctx context.Context) (*JudgeStats, error) {
	var total, allow, allowRisk, block, executed, success int64
	_ = r.db.WithContext(ctx).Raw(`SELECT COUNT(*) FROM "JudgeDecision"`).Scan(&total).Error
	_ = r.db.WithContext(ctx).Raw(`SELECT COUNT(*) FROM "JudgeDecision" WHERE "finalDecision" = 'ALLOW'`).Scan(&allow).Error
	_ = r.db.WithContext(ctx).Raw(`SELECT COUNT(*) FROM "JudgeDecision" WHERE "finalDecision" = 'ALLOW_WITH_RISK'`).Scan(&allowRisk).Error
	_ = r.db.WithContext(ctx).Raw(`SELECT COUNT(*) FROM "JudgeDecision" WHERE "finalDecision" = 'BLOCK'`).Scan(&block).Error
	_ = r.db.WithContext(ctx).Raw(`SELECT COUNT(*) FROM "JudgeDecision" WHERE "actualExecuted" = true`).Scan(&executed).Error
	_ = r.db.WithContext(ctx).Raw(`SELECT COUNT(*) FROM "JudgeDecision" WHERE "actualOutcome" = 'success'`).Scan(&success).Error
	rate := 0.0
	if executed > 0 {
		rate = float64(success) / float64(executed) * 100
	}
	return &JudgeStats{Total: total, Allow: allow, AllowRisk: allowRisk, Block: block, Executed: executed, Success: success, SuccessRate: rate}, nil
}

// SaveDecision is a convenience that accepts generic data (for stub compatibility).
func (r *JudgeRepository) SaveDecision(ctx context.Context, data interface{}) error {
	_ = data
	return nil
}
